ai/movie-recommendation/src/main/java/com/hazelcast/samples/ai/movies/TextSimilaritySearchImap.java

```diff
@@ -0,0 +1,364 @@
+
+        long start = Timer.nanos();
+        System.out.println("Loading " + movieIdToMeta.size() + " plot summaries");
+        movies.putAll(movieIdToMeta);
+        System.out.println("Loading " + movieIdToMeta.size() + " plot summaries "
+                + "took " + Timer.millisElapsed(start) + " milliseconds");
+    }
+
+    /**
+     * Reads the movie & plot summary data
+     */
+    public static void preparePlotSummaries(int limit) {
+        Scanner scanner = new Scanner(TextSimilaritySearchImap.class.getResourceAsStream("plot_summaries.txt"));
+        Pattern pattern = Pattern.compile("(\\d+)\t(.*)\n");
+        scanner.findAll(pattern)
+                .forEach(matchResult -> {
+                    String id = matchResult.group(1);
+                    String plot = matchResult.group(2);
+                    movieIdToPlotSummary.put(id, plot);
+                });
+        scanner = new Scanner(TextSimilaritySearchImap.class.getResourceAsStream("movie.metadata.tsv"));
+        pattern = Pattern.compile("(\\d+)\t[^\t]*\t([^\t]*)\t([^\t]*)\t.*\n");
+        scanner.findAll(pattern)
+                .limit(limit)
+                .forEach(matchResult -> {
+                    String id = matchResult.group(1);
+                    String name = matchResult.group(2);
+                    String releaseDate = matchResult.group(3);
+                    movieIdToMeta.put(id, new MovieMetadata(name, releaseDate, movieIdToPlotSummary.get(id)));
+                });
+    }
+    //endregion
+
+    public static void generateEmbeddings(IMap<String, MovieMetadata> movies) {
+        long start = Timer.nanos();
+        System.out.println("Generating embeddings for " + movies.size() + " plot summaries");
+        int embeddings;
+
+        if (offloaded) {
+            // offloading is not supported in executeOnEntries, so in order not to block the partition thread
+            // the invocations have to be performed one by one.
+            embeddings = movies.keySet().parallelStream()
+                    .map(movieId -> movies.executeOnKey(movieId, new MovieEmbeddingEntryProcessor()))
+                    .mapToInt(i -> i).sum();
+        } else {
+            // simpler version without offloading that can block partition thread for a long time
+            embeddings = movies.executeOnEntries(entry -> {
+                MovieMetadata value = entry.getValue();
+                if (value.plotSummary != null) {
+                    value.vector = new AllMiniLmL6V2EmbeddingModel().embed(value.plotSummary).content().vector();
+                    entry.setValue(value);
+                    return 1;
+                }
+                return 0;
+            }).values().stream().mapToInt(i -> i).sum();
+        }
+
+        System.out.println("Generating embeddings for " + embeddings + " plot summaries "
+                + "took " + Timer.secondsElapsed(start) + " seconds");
+    }
+
+    public static class MovieMetadata implements Serializable {
```

Comment:
record?

Reply:
`vector` field is mutable. this could be avoided, in this case it is more a matter of taste. both solutions would be equally simple